File: mcp/src/utils/replay.ts
import type { AuthCredential } from "../client.js";

export interface ReplayScriptInput {
  path: string;
  method: string;
  operation: Record<string, unknown>;
  target: string;
  vault: AuthCredential[];
  specVault?: unknown;
}

export function vaultHeadersFromSpec(specVault: unknown): Record<string, string> {
  const headers: Record<string, string> = {};
  if (!Array.isArray(specVault)) {
    return headers;
  }

  for (const entry of specVault) {
    if (
      entry &&
      typeof entry === "object" &&
      "header_name" in entry &&
      "token_value" in entry &&
      typeof entry.header_name === "string" &&
      typeof entry.token_value === "string"
    ) {
      headers[entry.header_name] = entry.token_value;
    }
  }

  return headers;
}

export function resolveVaultHeaders(
  vault: AuthCredential[],
  specVault?: unknown,
): Record<string, string> {
  const fromSpec = vaultHeadersFromSpec(specVault);
  if (Object.keys(fromSpec).length > 0) {
    return fromSpec;
  }

  const headers: Record<string, string> = {};
  for (const credential of vault) {
    if (credential.header_name && credential.token_value) {
      headers[credential.header_name] = credential.token_value;
    }
  }
  return headers;
}

export function buildReplayUrl(path: string, target: string): string {
  const host = target.split(",")[0]?.trim() || "target-domain.com";
  const base = host.startsWith("http") ? host : `https://${host}`;
  return `${base.replace(/\/$/, "")}${path}`;
}

export function generatePythonReplayScript(input: ReplayScriptInput): string {
  const method = input.method.toUpperCase();
  const vaultHeaders = resolveVaultHeaders(input.vault, input.specVault);
  const headers = {
    "User-Agent": "ShadowSchema-Replay/1.0",
    ...vaultHeaders,
  };

  const url = buildReplayUrl(input.path, input.target);
  let script = `import requests\nimport json\n\nurl = "${url}"\n\nheaders = ${JSON.stringify(headers, null, 4)}\n\n`;

  let payloadKwarg = "";
  const payload = input.operation["x-last-payload"];
  if (["POST", "PUT", "PATCH"].includes(method) && payload !== undefined) {
    script += `payload = ${JSON.stringify(payload, null, 4)}\n\n`;
    payloadKwarg = ", json=payload";
  }

  const vaultNote =
    Object.keys(vaultHeaders).length > 0
      ? `# Auth headers auto-injected from ShadowSchema Auth Vault\n`
      : `# No Auth Vault credentials captured yet for this session\n`;

  script = vaultNote + script;
  script += `response = requests.request("${method}", url, headers=headers${payloadKwarg})\n\nprint(f"Status: {response.status_code}")\nprint(response.text)\n`;
  return script;
}

File: internal/spec/replay.go
package spec

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

type replayRequest struct {
	Path   string `json:"path"`
	Method string `json:"method"`
}

func (s *SpecManager) mountReplayRoute(mux *http.ServeMux) {
	mux.HandleFunc("/export-replay", func(w http.ResponseWriter, r *http.Request) {
		enableCORS(w)
		if r.Method == http.MethodOptions {
			w.WriteHeader(http.StatusOK)
			return
		}

		var req replayRequest
		switch r.Method {
		case http.MethodGet:
			req.Path = r.URL.Query().Get("path")
			req.Method = r.URL.Query().Get("method")
		case http.MethodPost:
			if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
				http.Error(w, "Bad Request", http.StatusBadRequest)
				return
			}
		default:
			http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
			return
		}

		if req.Path == "" || req.Method == "" {
			http.Error(w, "path and method required", http.StatusBadRequest)
			return
		}

		sessionID, explicit, err := parseSessionIDQuery(r.URL.Query().Get("session_id"))
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		view, err := s.sessionReadView(sessionID, explicit)
		if err != nil {
			http.Error(w, err.Error(), http.StatusNotFound)
			return
		}

		detail, ok := endpointDetailFromDoc(view.Doc, req.Path)
		if !ok {
			http.Error(w, "endpoint not found", http.StatusNotFound)
			return
		}

		methodKey := strings.ToLower(req.Method)
		operations, ok := detail["operations"].(map[string]interface{})
		if !ok || operations[methodKey] == nil {
			http.Error(w, "method not found for endpoint", http.StatusNotFound)
			return
		}

		operation, ok := operations[methodKey].(map[string]interface{})
		if !ok {
			http.Error(w, "invalid operation payload", http.StatusInternalServerError)
			return
		}

		credentials, _ := s.listVaultCredentialsForSession(view.SessionID)
		script := buildPythonReplayScript(req.Path, strings.ToUpper(req.Method), operation, view.Target, credentials)

		w.Header().Set("Content-Type", "text/x-python; charset=utf-8")
		_, _ = w.Write([]byte(script))
	})
}

func buildPythonReplayScript(path, method string, operation map[string]interface{}, target string, credentials []AuthCredential) string {
	headers := map[string]string{
		"User-Agent": "ShadowSchema-Replay/1.0",
	}
	for _, credential := range credentials {
		if credential.HeaderName != "" && credential.TokenValue != "" {
			headers[credential.HeaderName] = credential.TokenValue
		}
	}

	url := buildReplayURL(path, target)
	headersJSON, _ := json.MarshalIndent(headers, "", "    ")

	var b strings.Builder
	if len(credentials) > 0 {
		b.WriteString("# Auth headers auto-injected from ShadowSchema Auth Vault\n")
	} else {
		b.WriteString("# No Auth Vault credentials captured yet for this session\n")
	}

	b.WriteString("import requests\nimport json\n\n")
	b.WriteString(fmt.Sprintf("url = %q\n\n", url))
	b.WriteString(fmt.Sprintf("headers = %s\n\n", string(headersJSON)))

	payloadKwarg := ""
	if payload, ok := operation["x-last-payload"]; ok && (method == "POST" || method == "PUT" || method == "PATCH") {
		payloadJSON, _ := json.MarshalIndent(payload, "", "    ")
		b.WriteString(fmt.Sprintf("payload = %s\n\n", string(payloadJSON)))
		payloadKwarg = ", json=payload"
	}

	b.WriteString(fmt.Sprintf("response = requests.request(%q, url, headers=headers%s)\n\n", method, payloadKwarg))
	b.WriteString("print(f\"Status: {response.status_code}\")\n")
	b.WriteString("print(response.text)\n")
	return b.String()
}

func buildReplayURL(path, target string) string {
	host := strings.TrimSpace(strings.Split(target, ",")[0])
	if host == "" {
		host = "target-domain.com"
	}
	if !strings.HasPrefix(host, "http://") && !strings.HasPrefix(host, "https://") {
		host = "https://" + host
	}
	return strings.TrimRight(host, "/") + path
}